crates/primitives/src/transaction/mod.rs

```diff
@@ -727,42 +692,50 @@ impl TransactionSigned {
     /// Inner encoding function that is used for both rlp [`Encodable`] trait and for calculating
     /// hash that for eip2728 does not require rlp header
     pub(crate) fn encode_inner(&self, out: &mut dyn bytes::BufMut, with_header: bool) {
-        if let Transaction::Legacy(TxLegacy { chain_id, .. }) = self.transaction {
-            let header = Header { list: true, payload_length: self.payload_len() };
-            header.encode(out);
-            self.transaction.encode_fields(out);
-
-            if let Some(id) = chain_id {
-                self.signature.encode_eip155_inner(out, id);
-            } else {
-                // if the transaction has no chain id then it is a pre-EIP-155 transaction
-                self.signature.encode_inner_legacy(out);
-            }
-        } else {
-            if with_header {
-                let header = Header { list: false, payload_length: self.payload_len() };
+        match self.transaction {
+            Transaction::Legacy(TxLegacy { chain_id, .. }) => {
+                // do nothing w/ with_header
+                let payload_length =
+                    self.transaction.fields_len() + self.signature.payload_len_legacy(chain_id);
+                let header = Header { list: true, payload_length };
                 header.encode(out);
+                self.transaction.encode_fields(out);
+                self.signature.encode_legacy(out, chain_id);
```

Comment:
Please note that we use `signature.encode_legacy()` at here and we use `signatue.encode()` at below.